### PR TITLE
Ransomware payload refactor

### LINKED ISSUE #1310: Ransomware Payload Refactor
Ransomware Payload has too many responsibilities. Break it out into focused components and use dependency injection to control the behavior of the ransomware payload.

- [ ] Transform BitFlipEncryptor into InPlaceEncryptor and inject it into RansomwarePayload's constructor (0.25d)
    - [x] Fix unit tests
- [ ] Put README logic into ReadmeHandler and inject it into RansomwarePayload's constructor (0.25d) @mssalvatore 
    - [x] Fix unit tests
- [x] Put inject select_production_safe_target_filesr() into RansomwarePayload's constructor (0d) @mssalvatore 
    - [x] Fix unit tests

### PR DESCRIPTION
**Pull request by mssalvatore**
# What does this PR do? 

Refactors the RansomwarePayload to limit its responsibilities and make it easier to change.
Resolves #1310

Add any further explanations here. 

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [x] Is the TravisCI build passing? 
* [ ] ~~Was the CHANGELOG.md updated to reflect the changes?~~
* [ ] ~~Was the documentation framework updated to reflect the changes?~~

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Have you successfully tested your changes locally? Elaborate:
    > Tested by running locally and running unit tests
* [ ] ~~If applicable, add screenshots or log transcripts of the feature working~~

### CODE REVIEW

File: monkey/infection_monkey/ransomware/ransomware_payload_builder.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+from infection_monkey.telemetry.messengers.legacy_telemetry_messenger_adapter import (
+    LegacyTelemetryMessengerAdapter,
+)
+from infection_monkey.utils.bit_manipulators import flip_bits
+
+EXTENSION = ".m0nk3y"
+CHUNK_SIZE = 4096 * 24
+
+
+def build_ransomware_payload(config: dict):
```
Comment (VakarisZ) [inline monkey/infection_monkey/ransomware/ransomware_payload_builder.py:18]: This looks more like a constructor. I'm not sure how would I build a different payload with this. I think this method should take a class/module as a parameter and use `_build_file_encryptor`, `_build_file_selector` and `_build_telemetry_messenger` defined in that class to create the payload object.

File: monkey/infection_monkey/ransomware/readme_utils.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import logging
```
Comment (VakarisZ) [inline monkey/infection_monkey/ransomware/readme_utils.py:1]: It's not the utility functions used by the readme feature, it is the readme feature itself. I'd rename this to `readme_dropper` or something

File: monkey/infection_monkey/ransomware/ransomware_payload_builder.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+
+def _build_file_encryptor():
+    return InPlaceFileEncryptor(
+        encrypt_bytes=flip_bits, new_file_extension=EXTENSION, chunk_size=CHUNK_SIZE
+    )
+
+
+def _build_file_selector():
+    targeted_file_extensions = TARGETED_FILE_EXTENSIONS.copy()
```
Comment (VakarisZ) [inline monkey/infection_monkey/ransomware/ransomware_payload_builder.py:35]: Why copy?

File: monkey/tests/unit_tests/infection_monkey/ransomware/test_ransomware_payload.py [hunk 4]
```diff
@@ -51,15 +34,46 @@ def ransomware_payload(build_ransomware_payload, ransomware_payload_config):
 
 
 @pytest.fixture
-def build_ransomware_payload(telemetry_messenger_spy):
+def build_ransomware_payload(
```
Comment (VakarisZ) [inline monkey/tests/unit_tests/infection_monkey/ransomware/test_ransomware_payload.py:37]: This whole file looks like a "mockery". What's the point of unit tests if we only assert that something got called? Aren't these kind of tests just tying everything together and making it harder to refactor, without actually testing any logic?

**Review by VakarisZ — APPROVED** [review]
Looks like good changes overall, just have some questions

File: monkey/tests/unit_tests/infection_monkey/ransomware/test_file_selectors.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+
+@pytest.fixture
+def file_selector():
+    return ProductionSafeTargetFileSelector(TARGETED_FILE_EXTENSIONS)
+
+
+def test_select_targeted_files_only(ransomware_test_data, file_selector):
+    selected_files = file_selector(ransomware_test_data)
+    print(ransomware_test_data)
```
[the remark this answers is not in the document]
Comment (mssalvatore) [reply monkey/tests/unit_tests/infection_monkey/ransomware/test_file_selectors.py:26]: Good catch.

File: monkey/tests/unit_tests/infection_monkey/ransomware/test_ransomware_payload.py (hunk 4 above)
Comment (mssalvatore) [reply monkey/tests/unit_tests/infection_monkey/ransomware/test_ransomware_payload.py:24]: It does test some logic. For example, it tests that encryption and readme actions are only performed if they are enabled in the configuration. It also tests that these actions are only performed if a target directory was specified. In addition, it tests that telemetries are constructed with the correct information.

There are more examples, but those are just a few.

**Pull request merged**